### PR TITLE
[Feature] Change Password API

### LINKED ISSUE #272: [Feature] Change Password API
- [x] New endpoint `/api/v1/company-user/change-password`
  - [ ] tests
Endpoint signature:
URL
`/api/v1/company-user/change-password`
Params
none
Query Params
`currentPassword` => current user's password
`newPassword` => new password

---
- [x] New endpoint `/api/v1/employee/change-password`
  - [ ] tests

Endpoint signature:
URL
`/api/v1/employee/change-password`
Params
none
Query Params
`currentPassword` => current user's password
`newPassword` => new password

---
Ensure that:
- [x] You are changing the password of the *current session's user*
- [x] Password field changes rely on middleware. Don't use Model#update.
- [x] User is activated
- [x] Current password matches

### PR DESCRIPTION
**Pull request by JavaLavaMT**
closes #272

close with associatedemployers/quartz#93
close with associatedemployers/granite#730

### CODE REVIEW

File: api/company-user/test.js [hunk 1]
```diff
@@ -160,4 +160,66 @@ describe('Acceptance :: Route :: Company User Utilities', () => {
...
+        newPassword:     '1234',
+        currentPassword: '1234'
+      };
+      let user = await (new CompanyUser({
+        email: 'test@aehr.org',
+        name:  {
+          first: 'Bob',
+          last:  'Ross'
+        },
+        password: '1234'
```
Comment (James1x0) [inline api/company-user/test.js]: Make this test fail by changing this, expecting "wrong password" in the response.

File: api/company-user/index.js [hunk 2]
```diff
@@ -41,6 +41,41 @@ exports.activationStatus = async (ctx) => {
   };
 };
 
+exports.changePassword = async (ctx) => {
+  let {
+    passwords, userId
+  } = ctx.request.body;
```
Comment (James1x0) [inline api/company-user/index.js]:
```suggestion
const { newPassword, currentPassword } = ctx.request.body;
```

File: api/company-user/index.js [hunk 3]
```diff
@@ -41,6 +41,41 @@ exports.activationStatus = async (ctx) => {
...
+
+  let user = await CompanyUser.findOne({
```
Comment (James1x0) [inline api/company-user/index.js]:
```suggestion
  let user = ctx.user;
```

File: api/company-user/index.js [hunk 4]
```diff
@@ -41,6 +41,41 @@ exports.activationStatus = async (ctx) => {
...
+        status: 400
+      }]
+    };
+    return;
+  }
+
+  user.password = passwords.newPassword;
+  await user.save();
+
+  ctx.status = 200;
```
Comment (James1x0) [inline api/company-user/index.js]:
```suggestion
  ctx.status = 204;
```

File: api/company-user/index.js [hunk 5]
```diff
@@ -41,6 +41,41 @@ exports.activationStatus = async (ctx) => {
...
+  ctx.body = {
+    companyUser: {
+      name: user.name,
+      _id:  user._id
+    }
+  };
+};
```
Comment (James1x0) [inline api/company-user/index.js]: We don't need a body

File: api/employee/index.js [hunk 6]
```diff
@@ -588,3 +588,38 @@ exports.suggestedDocuments = async (ctx) => {
   ctx.status = 200;
   ctx.body = suggestedDocuments;
 };
+
+exports.changePassword = async (ctx) => {
```
Comment (James1x0) [inline api/employee/index.js]:
```suggestion
exports.changePassword = require('../company-user').changePassword;
```

File: api/employee/test.js [hunk 7]
```diff
@@ -333,4 +333,66 @@ describe('Acceptance :: Route :: Employee', () => {
...
+      expect(res).to.have.status(200);
+
+      let updatedEmployee = await Employee.findById(employee._id);
+
+      const newPasswordChanged = await updatedEmployee.compareHash(passwords.newPassword);
+
+      expect(newPasswordChanged).to.equal(true);
+    });
+
+    it('should return a 400 when password cant be changed', async () => {
```
Comment (James1x0) [inline api/employee/test.js]: Make sure this test asserts currentpassword qp is checked against the current user's pass

File: api/company-user/index.js [hunk 8]
```diff
@@ -41,6 +41,41 @@ exports.activationStatus = async (ctx) => {
...
+  if (newPasswordMatchesCurrent) {
+    ctx.status = 400;
+    ctx.body = {
+      errors: [{
+        detail: 'New password cant be the same',
+        status: 400
+      }]
+    };
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/company-user/index.js (hunk 2 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/company-user/index.js (hunk 3 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/company-user/index.js (hunk 5 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/company-user/index.js (hunk 4 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/employee/index.js (hunk 6 above)
Comment (JavaLavaMT) [reply api/employee/index.js]: 🏁

File: api/company-user/test.js (hunk 1 above)
Comment (JavaLavaMT) [reply api/company-user/test.js]: 🏁

File: api/employee/test.js (hunk 7 above)
Comment (JavaLavaMT) [reply api/employee/test.js]: 🏁

Comment (JavaLavaMT) [reply api/employee/test.js]: 🏁

File: api/company-user/index.js [hunk 9]
```diff
@@ -41,6 +41,32 @@ exports.activationStatus = async (ctx) => {
...
+exports.changePassword = async (ctx) => {
+  const {
+    newPassword, currentPassword
+  } = ctx.request.body.passwords;
+
+  let { user } = ctx;
+
+  const PasswordMatchesCurrent = await user.compareHash(currentPassword);
+
+  if (!PasswordMatchesCurrent) {
```
Comment (James1x0) [inline api/company-user/index.js]:
```suggestion
  if (!await user.compareHash(currentPassword)) {
```
No need for this var in async land.

File: api/company-user/index.js [hunk 10]
```diff
@@ -41,6 +41,32 @@ exports.activationStatus = async (ctx) => {
   };
 };
 
+exports.changePassword = async (ctx) => {
+  const {
+    newPassword, currentPassword
+  } = ctx.request.body.passwords;
+
+  let { user } = ctx;
```
Comment (James1x0) [inline api/company-user/index.js]:
```suggestion
  const { user } = ctx;
```
Always use const unless you reassign the var.

File: api/company-user/index.js [hunk 11]
```diff
@@ -41,6 +41,32 @@ exports.activationStatus = async (ctx) => {
   };
 };
 
+exports.changePassword = async (ctx) => {
+  const {
+    newPassword, currentPassword
+  } = ctx.request.body.passwords;
```
Comment (James1x0) [inline api/company-user/index.js]: Why is this still `.passwords`??? 😞

File: api/company-user/test.js [hunk 12]
```diff
@@ -160,4 +160,65 @@ describe('Acceptance :: Route :: Company User Utilities', () => {
...
+      expect(res).to.have.status(204);
+
+      let updatedUser = await CompanyUser.findById(user._id);
+
+      const newPasswordChanged = await updatedUser.compareHash(passwords.newPassword);
+
+      expect(newPasswordChanged).to.equal(true);
+    });
+
+    it('should return a 400', async () => {
```
Comment (James1x0) [inline api/company-user/test.js]:
```suggestion
    it('should return a 400 when a current password is invalid', async () => {
```
Please be more descriptive if it seems vague.

File: api/company-user/test.js [hunk 13]
```diff
@@ -160,4 +160,65 @@ describe('Acceptance :: Route :: Company User Utilities', () => {
...
+        },
+        password: '1234'
+      })).save();
+
+      try {
+        await chai.request(app)
+        .post(`${baseUrl}`)
+        .set('X-Test-User', user._id.toString())
+        .send({
+          passwords
```
Comment (James1x0) [inline api/company-user/test.js]: Get rid of the passwords object like previously requested.

File: api/employee/config.json [hunk 14]
```diff
@@ -22,12 +22,17 @@
       "middleware": "uploadPicture",
       "permission": "create employee"
     },
+    "POST /employee/change-password": {
+      "secure": true,
+      "middleware": "changePassword",
+      "permission" : "create changes"
```
Comment (James1x0) [inline api/employee/config.json]: No permission plz

File: api/employee/config.json [hunk 15]
```diff
@@ -22,12 +22,17 @@
...
+    },
     "GET /employee/census/:id": {
       "secure": true,
       "middleware": "getCensusUpload",
       "permission": "create employee"
     },
-    "GET /employee/:id/resend-activation": {
+    "GET /employee/:id/resend-activation": {
```
Comment (James1x0) [inline api/employee/config.json]:
```suggestion
    "GET /employee/:id/resend-activation": {
```
Pointless change

File: api/employee/test.js [hunk 16]
```diff
@@ -333,4 +333,70 @@ describe('Acceptance :: Route :: Employee', () => {
...
+          last:  'Ross'
+        },
+        password: passwords.currentPassword
+      })).save();
+
+      let res = await chai.request(app)
+      .post(`${baseUrl}`)
+      .set('X-Test-User', employee._id.toString())
+      .send({
+        passwords
```
Comment (James1x0) [inline api/employee/test.js]: ✂️

File: api/company-user/index.js (hunk 9 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: good call

File: api/company-user/index.js (hunk 10 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/company-user/index.js (hunk 11 above)
Comment (JavaLavaMT) [reply api/company-user/index.js]: 🏁

File: api/employee/config.json (hunk 14 above)
Comment (JavaLavaMT) [reply api/employee/config.json]: 🏁

File: api/company-user/config.json [hunk 17]
```diff
@@ -4,6 +4,11 @@
   "secure": false,
   "routes": {
     "GET /company-user/activation-status/:user_id/:activation_id": "activationStatus",
-    "POST /company-user/activate/:user_id": "activate"
+    "POST /company-user/activate/:user_id": "activate",
+    "POST /company-user/change-password": {
+      "secure": true,
+      "middleware": "changePassword",
+      "permission": "update company users"
```
Comment (James1x0) [inline api/company-user/config.json]:
```suggestion
      "skipPermissions": true
```

**Pull request merged**